text/0000-nested-method-calls.md

```diff
@@ -0,0 +1,824 @@
+the reservation period, and hence no error results. This is because a
+reservation is equivalent to a shared borrow, and multiple shared
+borrows are allowed.
+
+Next, let's consider the sneaky example, where the argument attempts
+to mutate the vector that is being used in the receiver:
+
+```rust
+let mut v: Vec<String> = vec![format!("Hello, ")];
+let s: String = format!("foo");
+v[0].push_str({ v.push(s); "World!" });
+//              ^^^^^^^^^ sneaky attempt to mutate `v`
+```
+
+In this case, if we examine the resulting MIR, we can see that the
+borrow of `v` is almost **immediately** used, as part of the
+`IndexMut` operation:
+
+```rust
+// First evaluate `v[0]` to get a `&mut String`:
+tmp0 = &mut2 v;
+tmp1 = IndexMut::index_mut(tmp0, 0); // tmp0 used here!
+tmp2 = tmp1;
+
+// Next, evaluate `{ v.push(s); "World!" }` block:
+tmp3 = &mut2 v; // <-- Error! mutable borrow of `v` is active.
+... // see above
+```
+
+This implies that the mutable borrow will be active later on, when `v`
+is borrowed again during the arguments, and hence an error is still
+reported.
+
+Note that this same treatment will also rule out some "harmless"
+examples, such as this one:
+
+```rust
+v[0].push_str(&format!("{}", v.len()));
+```
+
+This might seem analogous to example 1, but in this case the mutable
+borrow of `v` is "activated" by the indexing, and hence `v` is
+considered mutably borrowed when `v.len()` is called, not reserved,
+which results in an error.
+
+# Detailed design
+[design]: #detailed-design
+
+### New MIR form for two-phase borrows
+
+Currently, the MIR rvalue for borrows has one of three forms (these
+are internal syntax only, naturally, since MIR doesn't have a defined
+written representation)
+
+    &'a <lvalue>
+    &'a mut <lvalue>
+    &'a unique <lvalue>
+    
+In either case, the rvalue returns a reference with lvalue `'a` that
+refers to the address of `lvalue` (an `lvalue` is a path that leads to
+memory). This can be either a shared, mutable, or unique reference
+(unique references are an internal concept that appears only in MIR;
+they are used when desugaring closures, but there is no direct
+equivalent in Rust surface syntax).
+
+This RFC proposes adding a third form: `&'a mut2 <lvalue>`. Like
+`&unique` borrows, this would be used by the compiler when desugaring
+and would not have a direct user representation for the time
+being. For most purposes, an `&mut2` borrow would act precisely the
+same as an `&mut` borrow; the borrow checker however would treat it
+differently, as described below.
+
+### When are two-phase borrows used
+
+Two-phase borrows would be used in the specific case of desugaring a
+call to an `&mut self` method. Currently, in the initially generated
+MIR, calls to such methods *always* have a "auto-mut-ref" inserted
+(this is because `vec.push()`, where `vec: &mut Vec<i32>`, is
+considered a *borrow* of `vec`, not a move). This "auto-mut-ref" will
+be changed from an `&mut` to an `&mut2`.
+
```

Comment:
That's not entirely correct:

```Rust
fn main() {
    let v = &mut vec![];
    v.push(0);
    v.push(1);
}
```

ATM we do not emit a reborrow:
```
        _1 = &mut _2; // _1 is `v`
        StorageLive(_4);
        _4 = _1;
        _3 = const <std::vec::Vec<T>>::push(_4, const 0i32) -> [return: bb4, unwind: bb3];
```

That is a bug, however, because if we don't emit a reborrow the code will not borrow-check (because of the multiple use of `v`).